core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1257,7 +1307,8 @@ class ReplicaManager(val config: KafkaConfig,
 
   def maybeUpdateMetadataCache(correlationId: Int, updateMetadataRequest: UpdateMetadataRequest) : Seq[TopicPartition] =  {
     replicaStateChangeLock synchronized {
-      if(updateMetadataRequest.controllerEpoch < controllerEpoch) {
+      confirmNotDeferringMetadataUpdatesWithZooKeeper()
```

Comment:
I think the check here and in `becomeLeaderOrFollower` is overkill. We already ensure that `beginMetadataChangeDeferral` cannot be called when using zk.